### ISSUE TITLE
Debugging Async methods not working

### ISSUE THREAD

**Issue by Kahbazi** [ArxOne/MrAdvice#91]
I install MrAdvice in my project and I have this code which I have no problem in debugging :

```
    class Program
    {
        static void Main()
        {
            Async().GetAwaiter().GetResult();
            Non_Async();
        }

        static async Task Async()
        {
            var foo = "bar";
            // can see foo in debug
        }

        static void Non_Async()
        {
            var foo = "bar";
            // can see foo in debug
        }
    }
```

when I add an Advice class to my code, I can no longer see local variables in async methods in debug mode.

```
    class Program
    {
        static void Main()
        {
            Async().GetAwaiter().GetResult();
            Non_Async();
        }

        static async Task Async()
        {
            var foo = "bar";
            // can NOT see foo in debug
        }

        static void Non_Async()
        {
            var foo = "bar";
            // can see foo in debug
        }
    }

   class AdviceA : Attribute, IMethodAdvice
    {
        public void Advise(MethodAdviceContext context)
        {
            Console.WriteLine("Entry From Advice A");
            context.Proceed();
            Console.WriteLine("Exit From Advice A");
        }
    }
```

**Comment by picrap** [comment]
This apparently related to `dnlib` (the library used to manipulate the assembly), I need to make some more tests to be sure. I'll let you know.

**Comment by Kahbazi** [comment]
thank you. there is another problem, I think it's related to this issue.
in debug whenever I `StepOver(F10) ` a statement which has `await` the result is like when I press `Continue(F5)` .

**Comment by royshea** [comment]
I'm working on a project that is also running into problems with debugging after adding MrAdvice.  The symptoms that @Kahbazi mentioned mirror what I am seeing.

**Comment by picrap** [comment]
Yes, this is related to the `dnlib` library, so unfortunately out of Mr. Advice scope. I'm waiting for a fix, after having tried unsuccessfully to fix it.

**Comment by shivamtripathi** [comment]
@picrap What about MrAdvice.Fody. As that one uses `Mono.Cecil` so this issue should not be there. But I checked the nuget package and found that `MrAdvice.Fody` is dependent upon `MrAdvice 2.0`. While testing I found that this issue persists there too though logically it shouldn't as it is not using `dnlib`. I know that `MrAdvice.Fody` is obsolete and is not being maintained but if it provides the basic functionality (Simple Advice etc) and doesn't have any major bug, I would like to use it in my project. By any chance can I get that? Thanks

**Comment by picrap** [comment]
Sure, you still can use the latest working version of `MrAdvice.Fody`, available at https://www.nuget.org/packages/MrAdvice.Fody/1.3.3

**Comment by shivamtripathi** [comment]
Thanks but just tried it and it is throwing the following exception on build - 
1>------ Rebuild All started: Project: Check, Configuration: Debug Any CPU ------
1>    Fody (version 1.28.3.0) Executing
1>MSBUILD : error : Fody: An unhandled exception occurred:
1>MSBUILD : error : Exception:
1>MSBUILD : error : Could not load file or assembly 'Mono.Cecil, Version=0.9.6.0, Culture=neutral, PublicKeyToken=0738eb9f132ed756' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
1>MSBUILD : error : StackTrace:
1>MSBUILD : error :    at System.Signature.GetSignature(Void* pCorSig, Int32 cCorSig, RuntimeFieldHandleInternal fieldHandle, IRuntimeMethodInfo methodHandle, RuntimeType declaringType)
1>MSBUILD : error :    at System.Reflection.RuntimeMethodInfo.FetchNonReturnParameters()
1>MSBUILD : error :    at System.Reflection.RuntimeMethodInfo.GetParametersNoCopy()
1>MSBUILD : error :    at System.Reflection.RuntimePropertyInfo.GetIndexParametersNoCopy()
1>MSBUILD : error :    at System.Reflection.RuntimePropertyInfo.GetIndexParameters()
1>MSBUILD : error :    at System.RuntimeType.GetPropertyCandidates(String name, BindingFlags bindingAttr, Type[] types, Boolean allowPrefixLookup)
1>MSBUILD : error :    at System.RuntimeType.GetPropertyImpl(String name, BindingFlags bindingAttr, Binder binder, Type returnType, Type[] types, ParameterModifier[] modifiers)
1>MSBUILD : error :    at System.Type.GetProperty(String name, BindingFlags bindingAttr, Binder binder, Type returnType, Type[] types, ParameterModifier[] modifiers)
1>MSBUILD : error :    at PropertyDelegateBuilder.BuildPropertySetDelegate[T](Type type, String propertyName) in c:\ConsoleBuildAgent\work\ed448661dbb30d2e\FodyIsolated\DelegateBuilders\PropertyDelegateBuilder.cs:line 9
1>MSBUILD : error :    at DelegateBuilder.BuildDelegateHolder(Type weaverType) in c:\ConsoleBuildAgent\work\ed448661dbb30d2e\FodyIsolated\DelegateBuilders\DelegateBuilder.cs:line 25
1>MSBUILD : error :    at DelegateBuilder.GetDelegateHolderFromCache(Type weaverType) in c:\ConsoleBuildAgent\work\ed448661dbb30d2e\FodyIsolated\DelegateBuilders\DelegateBuilder.cs:line 16
1>MSBUILD : error :    at InnerWeaver.InitialiseWeavers() in c:\ConsoleBuildAgent\work\ed448661dbb30d2e\FodyIsolated\InnerWeaver.cs:line 111
1>MSBUILD : error :    at InnerWeaver.Execute() in c:\ConsoleBuildAgent\work\ed448661dbb30d2e\FodyIsolated\InnerWeaver.cs:line 61
1>MSBUILD : error : Source:
1>MSBUILD : error : mscorlib
1>MSBUILD : error : TargetSite:
1>MSBUILD : error : Void GetSignature(Void*, Int32, System.RuntimeFieldHandleInternal, System.IRuntimeMethodInfo, System.RuntimeType)
1>MSBUILD : error : 
1>      Finished Fody 13ms.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========


I tried installing manually Mono.Cecil Version 0.9,6.0 but no luck. Not sure if it is Fody issue or MrAdvice.Fody.

**Comment by shivamtripathi** [comment]
Hi, 
I found the issue. It is Fody Version 1.28.03. I updated the "MrAdvice.tt" file to include the latest version of Fody (1.29.4) in "MrAdvice-Fody" branch and it is working fine now. The only issue is that how to get this updated version via Nuget. Can you help me by publishing an updated version of MrAdvice.Fody? I know it is not being maintained but my project can't go live with this missing feature of DnLib. We need the debugging to work properly. That's where MrAdvice.Fody comes handy provided it doesn't have any major issue.

Thanks, 
Shivam

**Comment by picrap** [comment]
If you submit a pull request to the correct branch ("fody") with a new version number (1.3.4), I'll just have to accept it and it is build and published automatically (you can track Mr Advice builds at https://ci.appveyor.com/project/picrap/mradvice/history)

**Comment by shivamtripathi** [comment]
Hi, I am not able to create a pull request due to permission issue. Could you please provide me rights? Thanks.

**Comment by picrap** [comment]
To submit a pull request, you need to:

1. fork the repository, 
2. work on your own fork (ensure to work on `fody` branch),
3. commit and push your changes to your own repository
4. then you can submit a pull request to original repository, which is mine.
Why all this? Because we don't really know each other, so I'd like to review your changes in a pull request before accepting them.

**Comment by shivamtripathi** [comment]
Got it. NP.

**Comment by shivamtripathi** [comment]
Done. Created a PR. Please have a look.

**Comment by picrap** [comment]
Merged, building. The new version should be available on NuGet soon.

**Comment by picrap** [comment]
The async debug was solved in dnlib, so MrAdvice 2.5.3 uses it and solves the problem too.

**Comment by Kahbazi** [comment]
@picrap thanks, really appreciate it.
